grails-app/views/replenishment/_printPage.gsp

```diff
@@ -0,0 +1,103 @@
+<%@ page import="org.apache.commons.lang.StringEscapeUtils" defaultCodec="html" %>
+
+<div class="page" style="page-break-after: ${pageBreakAfter};">
+    <table id="requisition-items" class="fs-repeat-header" border="0">
+        <thead style="display: table-row-group">
+            <tr class="">
+                <td colspan="11">
+                    <h4 class="title">${groupName}</h4>
+                </td>
+            </tr>
+            <tr class="theader">
+                <th><warehouse:message code="report.number.label"/></th>
+                <th>${warehouse.message(code: 'product.productCode.label')}</th>
+                <th>${warehouse.message(code: 'product.name.label')}</th>
+                <th class="center" style="min-width: 150px;">${warehouse.message(code: 'default.lotSerialNo.label')}</th>
+                <th class="center">${warehouse.message(code: 'orderItem.expiry.label')}</th>
+                <th class="center">${warehouse.message(code: 'orderItem.transferToBin.label')}</th>
+                <th class="center">${warehouse.message(code: 'orderItem.qtyToTransfer.label')}</th>
+                <th class="center">${warehouse.message(code: 'inventoryLevel.binLocation.label')}</th>
+                <th class="center">${warehouse.message(code: 'requisitionItem.suggestedPick.label')}</th>
+                <th class="center" style="min-width: 100px">${warehouse.message(code:'default.notes.label')}</th>
+            </tr>
+        </thead>
+        <tbody>
+            <g:unless test="${lineItems}">
+                <tr>
+                    <td colspan="10" class="middle center">
+                        <span class="fade">
+                            <warehouse:message code="default.none.label"/>
+                        </span>
+                    </td>
+                </tr>
+            </g:unless>
+            <g:each in="${lineItems}" status="i" var="lineItem">
+
+                <g:set var="picklistItems" value="${pickListItemsByOrder[lineItem.id]?.findAll { it.quantity > 0 }}"/>
```

Comment:
This is what I was trying to avoid. I want the GSP to be as dumb as possible, so all of this data manipulation and transforming done here and in the next GSP should be handled in the controller or service layer. 

The GSP should be as simple as 
```
<g:each var="picklistItem" in="${picklistItems}">
    <tr>
        <td>${picklistItem.column1}</td>
        <td>${picklistItem.column2}</td>
        <td>${picklistItem.column3}</td>
    </tr>
```
And all the complex logic of creating the picklist items would be handled in the controller / service. That way we could have a print picklist action anywhere we need it, the GSP would be the same for each and the controller / service would transform the data to fit into the buckets we created for the GSP (header, groups, items, items grouped by groups). 

This probably would have required us creating our own Picklist and PicklistItem command classes that were mapped more cleanly to the GSP. For example, the handling of the zones would just be done as a group. 
```
PicklistCommand
    GroupByCommand sections 
    Map<GroupByCommand, List<PicklistItemCommand> getPicklistItemMap
```
Sections should actually be another command object

```
List<PicklistItemCommand> picklistItems = transformPicklistItems(picklist.picklistItems)
Map<GroupByCommand, List<PicklistItemCommand> picklistItemMap = picklistItems.groupBy { new GroupByCommand(it.whatever) }
PicklistCommand new PicklistCommand(sections: picklistItemMap.keySet(), picklistItemMap: picklistItemMap)
```
In this case we want to break it down by zone. In another case we might want to group it by categories.

However, I'm realizing that what I'm asking for might be too difficult so I'll let @awalkowiak decide if this meets the ticket's acceptance criteria and we can move it to QA.

Reply:
The more I think about it, we could have gone even further and created generic reporting interfaces to deal with the elements of the report (i.e. Header, Groups, Rows) and then you could pass all of this data along with the column definitions as parameters to a generic "tabular report" GSP. Turning a very specific "print transfer order picklist" feature into a more generic "generate report" that is customizable based on the model / parameters passed to it.

Reply:
@jmiranda I think we'll go with as generic as we can, and then we are gonna create a tech debt/improvement ticket to add commands and to refactor the existing print (for outbound sm) to use this new approach.